dp-hello-world-event/event/consumer.go

```diff
@@ -0,0 +1,104 @@
+package event
+
+import (
+	"context"
+	"errors"
+
+	"github.com/ONSdigital/dp-hello-world-event/schema"
+	kafka "github.com/ONSdigital/dp-kafka/v2"
+	"github.com/ONSdigital/log.go/log"
+)
+
+//go:generate moq -out mock/handler.go -pkg mock . Handler
+
+// MessageConsumer provides a generic interface for consuming []byte messages
+type MessageConsumer interface {
+	Channels() *kafka.ConsumerGroupChannels
+}
+
+// Handler represents a handler for processing a single event.
+type Handler interface {
+	Handle(ctx context.Context, helloCalled *HelloCalled) error
+}
+
+// Consumer consumes event messages.
+type Consumer struct {
+	closing chan bool
+	closed  chan bool
+}
+
+// NewConsumer returns a new consumer instance.
+func NewConsumer() *Consumer {
+	return &Consumer{
+		closing: make(chan bool),
+		closed:  make(chan bool),
+	}
+}
+
+// Consume converts messages to event instances, and pass the event to the provided handler.
+func (consumer *Consumer) Consume(ctx context.Context, messageConsumer MessageConsumer, handler Handler) {
+
+	go func() {
+		defer close(consumer.closed)
+
+		for {
+			select {
+			case message := <-messageConsumer.Channels().Upstream:
+				messageCtx := context.Background()
+				processMessage(messageCtx, message, handler)
+				message.Release()
+			case <-consumer.closing:
+				log.Event(ctx, "closing event consumer loop", log.INFO)
+				return
+			}
+		}
+	}()
+
+}
+
+// Close safely closes the consumer and releases all resources
+func (consumer *Consumer) Close(ctx context.Context) (err error) {
+	if ctx == nil {
+		ctx = context.Background()
+	}
+
+	close(consumer.closing)
```

Comment:
if we implement the changes described above, here we would just need to close the kafka consumer (or stop listening to it). In fact, if we close the Kafka consumer from the service package, this function would not be needed at all (the consume loop would exit once the kafka consumer closes the upstrem channel, which happens during consumer Close(): https://github.com/ONSdigital/dp-kafka/blob/main/consumer_group.go#L194)